server/cluster_worker.go

```diff
@@ -22,59 +22,37 @@ import (
 	"github.com/pingcap/kvproto/pkg/pdpb"
 )
 
-func (c *RaftCluster) addReplicaBalanceOperator(region *metapb.Region, leader *metapb.Peer, downPeers []*pdpb.PeerStats) (*balanceOperator, error) {
-	if !c.balancerWorker.allowBalance() {
-		return nil, nil
-	}
-
-	balancer := newReplicaBalancer(region, leader, downPeers, &c.s.cfg.BalanceCfg)
-	_, balanceOperator, err := balancer.Balance(c.cachedCluster)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	if balanceOperator == nil {
-		return nil, nil
-	}
-
-	if c.balancerWorker.addBalanceOperator(balanceOperator.getRegionID(), balanceOperator) {
-		return balanceOperator, nil
-	}
-
-	// Em, the balance worker may have already added a BalanceOperator.
-	return c.balancerWorker.getBalanceOperator(region.GetId()), nil
-}
-
 func (c *RaftCluster) handleRegionHeartbeat(region *metapb.Region, leader *metapb.Peer, downPeers []*pdpb.PeerStats) (*pdpb.RegionHeartbeatResponse, error) {
 	// If the region peer count is 0, then we should not handle this.
 	if len(region.GetPeers()) == 0 {
 		log.Warnf("invalid region, zero region peer count - %v", region)
 		return nil, errors.Errorf("invalid region, zero region peer count - %v", region)
 	}
 
+	bw := c.balancerWorker
 	regionID := region.GetId()
-	balanceOperator := c.balancerWorker.getBalanceOperator(regionID)
-	var err error
-	if balanceOperator == nil {
-		balanceOperator, err = c.addReplicaBalanceOperator(region, leader, downPeers)
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
-		if balanceOperator == nil {
-			return nil, nil
-		}
+
+	err := bw.checkReplicas(region, leader, downPeers)
```

Comment:
If we check `allowBalance` here, the replica balance may be starved by other resource balance. However, we should check if TiKV can handle that later or we should limit the replica balance somehow.